### ISSUE TITLE
Format Breaking Changes Candidates

### ISSUE THREAD

**Issue by crepererum** [JDASoftwareGroup/kartothek#41]
This is a meta issue describing which changes we would like to make now but which are incompatible with the current format.

## Type Normalization
The following type normalizations are currently not implemented but could be:
- `decimal[P, S] -> decimal[max_p(S), S]`
- `date{32, 64} -> date64`
- `time{32, 64}[U] -> time64[U]`
- structs (nested normalization)

## Index Handling
Reject non-integer/range indices, use `reset_index` and drop index information before writing data. Always restore as normalized (`reset_index`) indices, even when applying predicates.

## Pandas-specific Metadata
Pandas-specific metadata is part of the Arrow schema but is not part of the Arrow Type system. It captures information like the index type. If *Index Handling* is implemented, we could drop the entire pandas metadata field. This would simplify interopt with other languages/frameworks.

**Comment by xhochy** [comment]
> `decimal[P, S] -> decimal[38, S]`

The limitation in Arrow is currently that it only supports what fits into an `int128`. This is where the magic 38 is coming from. This may change in future.

**Comment by fjetter** [comment]
The way I see it we don't necessarily need to *upcast* everything for the schema, do we? As long as there is a *fixed* width for a given type and our compatibility checks ensure that an entire type family is considered to be compatible our requirements should be satisfied.

**Comment by crepererum** [comment]
@fjetter yes. Currently, we upcast only during write and do not upcast the common metadata again during read. So adding additional upcast rules would break existing datasets, BUT: we could just upcast during common metadata read again (actually, we should do that for compatibility with other frameworks anyway). So the introduction would only break forward compatibility for older ktk versions.

**Comment by fjetter** [comment]
I'm not that concerned with breakage but I'm rather asking myself: Why do we upcast to the broadest width? Why not the smallest? What would we do if `pyarrow` introduces a `int96` or `int128`? Do we change our casting rules?
My point is: As long as we are deterministic, i.e. we set a fixed cast target, this shouldn't matter too much. The question is obviously: What should this _fixed_ type? At least to my understanding the current casting rules cast to the _most common_ type. Most common in the sense that these types are used by pandas as defaults for the given type family.

**Comment by crepererum** [comment]
> Why do we upcast to the broadest width? Why not the smallest? 

We use the largest because it the common metadata then describes a type that can hold all variables of all partitions (aka container type). That's the whole point of the type system documentation and also explains why ints cannot be packed into floats or the other way around.

> What would we do if pyarrow introduces a int96 or int128? Do we change our casting rules?

Depends. It would make sense to upcast to int128 in that case, but we might not want to that if this means that all libs break because they cannot handle this type (numpy for example). Or in other words (also as described in the type system docs): find a container type that still doesn't break the ecosystem.

>  Most common in the sense that these types are used by pandas as defaults for the given type family.

Pandas is exactly NOT a good blueprint for the type system (see float VS int discussion again). Most common means "the container / common type can hold all values of all types that are upcasted into that exact container / common type" (which is not the case for int<->float) and that "make semantically sense" (see discussion on why bools should not be upcasted to ints in the type system docs).

**Comment by lr4d** [comment]
If partition labels and partition indices (recently deprecated) are both removed, it will be nice for UX if `parse_input_to_metapartition` accepts what is currently the value of the `data` argument, given that `data` would be the only key remaining in the input dictionary.

That is, instead of the current:
```
dfs = [
    {
        "data": {
            "core-table": pd.DataFrame({"col1": ["x"]}),
            "aux-table": pd.DataFrame({"f": [1.1]}),
        }
    },
    {
        "data": {
            "core-table": pd.DataFrame({"col1": ["y"]}),
            "aux-table": pd.DataFrame({"f": [1.2]}),
        }
    },
]
```
Allow:
```
dfs = [
    {
        "core-table": pd.DataFrame({"col1": ["x"]}),
        "aux-table": pd.DataFrame({"f": [1.1]}),
    },
    {
        "core-table": pd.DataFrame({"col1": ["y"]}),
        "aux-table": pd.DataFrame({"f": [1.2]}),
    },
]
```

**Comment by lr4d** [comment]
I'd suggest we make the `kartothek.io` read/write/update function signatures more user-friendly and consistent:
Currently we have:
```
store_dataframes_as_dataset(
    store,
    dataset_uuid,
    dfs,
...
```
```
update_dataset_from_dataframes__iter(
    df_generator,
    store=None,
    dataset_uuid=None
...
```
The latter approach makes sense to me, as one should be able to pass a `dataset_factory` instead of `store` and `dataset_uuid` combination.
The order of positional arguments changing is also not very user-friendly, as, for a generic approach, it forces you to use keyword arguments and look at the name of the kwarg used to pass the dataframes to the function (`dfs`, `df_generator`, `ddf`....)

**Comment by crepererum** [comment]
@lr4d I agree with your proposal, this is not format-breaking (in terms of on-disk / in-store format / specification). It is API breaking. See https://kartothek.readthedocs.io/en/latest/versioning.html for details.
